common/src/main/java/com/google/auto/common/BasicAnnotationProcessor.java

```diff
@@ -255,69 +271,96 @@ private String processingErrorMessage(String target) {
+   * previously ill-formed elements.
+   *
+   * <p>Note that the elements deferred by processing steps are guaranteed to be well-formed;
+   * therefore, they are ignored (not returned) here, and they will be considered directly in the
+   * {@link #process(ImmutableSetMultimap)} method.
    */
-  private ImmutableSetMultimap<TypeElement, Element> validElements(RoundEnvironment roundEnv) {
-    ImmutableSet<ElementName> prevDeferredElementNames = ImmutableSet.copyOf(deferredElementNames);
-    deferredElementNames.clear();
+  private ImmutableSetMultimap<TypeElement, Element> getWellFormedElements(
+      RoundEnvironment roundEnv) {
+    ImmutableSet<PackageOrTypeElementBluePrint> prevIllFormedElementBluePrints =
+        ImmutableSet.copyOf(illFormedElementBluePrints);
+    illFormedElementBluePrints.clear();
 
-    ImmutableSetMultimap.Builder<TypeElement, Element> deferredElementsByAnnotationBuilder =
+    ImmutableSetMultimap.Builder<TypeElement, Element> readyDeferredElementsByAnnotationBuilder =
         ImmutableSetMultimap.builder();
-    for (ElementName deferredElementName : prevDeferredElementNames) {
-      Optional<? extends Element> deferredElement = deferredElementName.getElement(elements);
-      if (deferredElement.isPresent()) {
+    for (PackageOrTypeElementBluePrint deferredElementBluePrint : prevIllFormedElementBluePrints) {
+      Optional<? extends Element> deferredTypeElement =
+          deferredElementBluePrint.getElement(elementUtils);
+      if (deferredTypeElement.isPresent()) {
         findAnnotatedElements(
-            deferredElement.get(),
+            deferredTypeElement.get(),
             getSupportedAnnotationTypeElements(),
-            deferredElementsByAnnotationBuilder);
+            readyDeferredElementsByAnnotationBuilder);
       } else {
-        deferredElementNames.add(deferredElementName);
+        illFormedElementBluePrints.add(deferredElementBluePrint);
       }
     }
 
-    ImmutableSetMultimap<TypeElement, Element> deferredElementsByAnnotation =
-        deferredElementsByAnnotationBuilder.build();
+    ImmutableSetMultimap<TypeElement, Element> readyDeferredElementsByAnnotation =
+        readyDeferredElementsByAnnotationBuilder.build();
 
-    ImmutableSetMultimap.Builder<TypeElement, Element> validElements =
+    ImmutableSetMultimap.Builder<TypeElement, Element> validatedElementsBuilder =
         ImmutableSetMultimap.builder();
 
-    Set<ElementName> validElementNames = new LinkedHashSet<>();
+    // For optimization purposes, the PackageOrTypeElementBluePrint that have already
+    // been verified to be well-formed are stored.
+    Set<PackageOrTypeElementBluePrint> wellFormedPackageOrTypeElementBluePrints =
+        new LinkedHashSet<>();
 
-    // Look at the elements we've found and the new elements from this round and validate them.
+    /* Look at the found ready deferred elements and the new elements from this round
```

Comment:
I think this should be something like "the already-deferred elements"? But maybe I'm misunderstanding what "ready" means here.

Reply:
I've refined variable names and improved this comment for better readability and clarity of purpose.